diff --git a/HW7/src/cis5550/flame/FlameContextImpl.java b/HW7/src/cis5550/flame/FlameContextImpl.java
index 4f37d3f4..8893b92d 100644
--- a/HW7/src/cis5550/flame/FlameContextImpl.java
+++ b/HW7/src/cis5550/flame/FlameContextImpl.java
@@ -7,6 +7,7 @@ import cis5550.tools.Logger;
 import cis5550.tools.Partitioner;
 import cis5550.tools.Partitioner.Partition;
 import cis5550.tools.Serializer;
+import cis5550.tools.KeyEncoder;
 
 import java.io.IOException;
 import java.nio.file.Files;
@@ -98,10 +99,10 @@ public class FlameContextImpl implements FlameContext {
                     + "&kvs=" + getKVS().getCoordinator());
 
             if(fromKey != null) {
-                query.append("&from=").append(fromKey);
+                query.append("&from=").append(KeyEncoder.encode(fromKey));
             }
             if(toKeyExclusive != null) {
-                query.append("&to=").append(toKeyExclusive);
+                query.append("&to=").append(KeyEncoder.decode(toKeyExclusive));
             }
 
             queryParams.forEach((k, v) -> {
diff --git a/HW7/src/cis5550/flame/operation/CogroupOperation.java b/HW7/src/cis5550/flame/operation/CogroupOperation.java
index 9896798e..06212b91 100644
--- a/HW7/src/cis5550/flame/operation/CogroupOperation.java
+++ b/HW7/src/cis5550/flame/operation/CogroupOperation.java
@@ -27,7 +27,7 @@ public class CogroupOperation implements Operation {
         Map<String, String> map1 = buildLists(it1);
         Map<String, String> map2 = buildLists(it2);
 
-        Set<String> commonKeys = map1.keySet();
+        Set<String> commonKeys = new HashSet<>(map1.keySet());
         commonKeys.retainAll(map2.keySet());
 
         for(String key : commonKeys){
diff --git a/HW7/src/cis5550/flame/operation/OperationContext.java b/HW7/src/cis5550/flame/operation/OperationContext.java
index 0a29a39a..6e602b32 100644
--- a/HW7/src/cis5550/flame/operation/OperationContext.java
+++ b/HW7/src/cis5550/flame/operation/OperationContext.java
@@ -3,6 +3,7 @@ package cis5550.flame.operation;
 import cis5550.kvs.KVS;
 import cis5550.kvs.KVSClient;
 import cis5550.kvs.Row;
+import cis5550.tools.KeyEncoder;
 import cis5550.webserver.Request;
 import cis5550.tools.Logger;
 
@@ -46,11 +47,11 @@ public class OperationContext {
     }
 
     public void from(String from){
-        this.startFrom = from;
+        this.startFrom = from == null ? null : KeyEncoder.decode(from);
     }
 
     public void to(String to){
-        this.endExclusive = to;
+        this.endExclusive = to == null ? null : KeyEncoder.decode(to);
     }
 
     public String from(){
